### ISSUE TITLE
[core] XML attributes contain unprintable bytes

### ISSUE THREAD

**Issue by sergey-wowwow** [skylot/jadx#844]
APK: https://drive.google.com/file/d/1msaMxk32XHruCtEaQDJNk0IGnb7oOWBo/view?usp=sharing

Open ``` AndroidManifest.xml ``` and check attributes e.g. 
```xml
    <application android:theme="@style/" android:label="@string/" android:icon="@mipmap/i_launcher_app" android:name="com.drweb.p000s.App" android:persistent="true" android:allowBackup="false" android:hardwareAccelerated="true" android:supportsRtl="true" android:roundIcon="@mipmap/i_launcher_app_round" android:appCategory="7" android:appComponentFactory="androidx.core.app.CoreComponentFactory">
```

**Comment by sergey-wowwow** [comment]
Seems that those values are parsed in ``` /jadx-core/src/main/java/jadx/core/xmlgen/ResTableParser.java ```
```java
		String[] keyStrings = null;
		if (keyStringsOffset != 0) {
			is.skipToPos(keyStringsOffset, "Expected keyStrings string pool");
			keyStrings = parseStringPool(); // <<<
		}

		PackageChunk pkg = new PackageChunk(id, name, typeStrings, keyStrings);
```
That package is definitely obfuscated, @skylot do you have any ideas
1. how to restore valid values
2. at least where to filter original values? When written in file ``` BinaryXMLParser.java ```, or when they are decoded?

**Comment by sergey-wowwow** [comment]
@skylot 
Can you please suggest something here? I meet this bug pretty often, so I want to fix it asap

**Comment by jpstotz** [comment]
The resources are heavily obfuscated and the key-strings are meaningless - Android just ignores them. Android only uses the resourceIds when loading a resource string, hence the only chance I see here is to check if the resource key name is an obfuscated string like in this case the character 0x1 and then ignore the string and generate a key name from it's resourceId instead.

**Comment by sergey-wowwow** [comment]
@jpstotz 
Yea, that's why I asked for an advise where is the best place to modify code and generate own resource names

**Comment by jpstotz** [comment]
I made some tests and it seems like the best way to handle such situations is to correct the resource names in the resources.arsc reader. 

In detail I modified the end of the method `private static String extractString8(byte[] strArray, int offset)`
https://github.com/skylot/jadx/blob/master/jadx-core/src/main/java/jadx/core/xmlgen/CommonBinaryParser.java#L62:

		String s =  new String(arr, ParserStream.STRING_CHARSET_UTF8);
		if (s.length()==1) {
			s = "String_at_offset_0x"+Integer.toHexString(offset);
		}
		return s;

If you apply this hack all resources get generated unique and readable resource names that in the end are also referenced in the AndroidManifest.xml.

**Comment by jpstotz** [comment]
After some more tests I noticed that indeed you already pointed out a good position to perform the deobfuscation: https://github.com/skylot/jadx/blob/master/jadx-core/src/main/java/jadx/core/xmlgen/ResTableParser.java#L154

But back to your questions:

> 1. how to restore valid values

We are talking about Android resource names. If we just generate unique an ASCII name for each key this should be sufficient.

 > 2. at least where to filter original values? When written in file BinaryXMLParser.java, or when they are decoded?

The correct approach IMHO would be to scan the keyValues for duplicate strings. I have not checked it but I would assume that in non-obfuscated apps the keyValues array does not contain a single duplicate value. Hence we could scan this array for duplicate and non-printable values and generate a new key value instead. The following code is a very simplified version without the check for duplicate values. 

```
		keyStrings = parseStringPool();
		for (int i=0; i< keyStrings.length; i++) {
			String key = keyStrings[i];
			if (key.length()==1 && key.charAt(0)<32) { // ASCII 32 = SPACE
				keyStrings[i] = String.format("jadx_generated_key_%d", i);
			}
		}

```

**Comment by skylot** [comment]
Sorry guys I can't suggest anything here because I don't know much about resource processing in Android. I only can your implement suggestions :)
In PR #909 I implement @jpstotz approach, please check and review, hope this will help.

**Comment by skylot** [comment]
Fixed in PR #909
